StatefulSet scale subresource

crimsonfaith91 opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/devel/pull-requests.md#the-pr-submit-process and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/devel/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/devel/pull-requests.md#write-release-notes-if-needed
-->

**What this PR does / why we need it**: This PR implements scale subresource for StatefulSet.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #46005

**Special notes for your reviewer**:

**Release note**:
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access)
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`.
-->
```release-note
StatefulSet uses scale subresource when scaling in accord with ReplicationController, ReplicaSet, and Deployment implementations.
```
**Feature Checklist**:
- [x] Introduce Registry interface for storage purpose
- [x] Introduce `ScaleREST New(), Get() and Update()` utility functions
- [x] Create a `ScaleREST` object at `NewREST()` and return it
- [x] Enable scale subresource by adding `/scale` field to the storage map

**Testing Checklist**:
- Unit testing
  - [x] Modify `newStorage()` to call `NewStorage()`, and change all unit tests accordingly
  - [x] Add unit tests for `ScaleREST Get() and Update()` utility functions
  - [x] Add missing unit test for `ShortNames`

- Manual testing
  - [x] Verify existence of the subresource using `kubectl proxy` command
  - [x] Modify the subresource using `curl` via `POST`

- e2e testing
  - [x] Add e2e tests using `RESTClient`

crimsonfaith91 commented:
/unassign @smarterclayton @gmarek

crimsonfaith91 committed:
implement statefulset scale subresource

crimsonfaith91 commented:
/assign @janetkuo @enisoc 
i shifted the change to v1beta2.

crimsonfaith91 commented:
/unasign @soltysh @yifan-gu

crimsonfaith91 commented:
/unassign @soltysh @yifan-gu

crimsonfaith91 commented:
/test pull-kubernetes-unit

enisoc commented:
Looks like a great start! It just needs integration/e2e tests to show that the subresource is properly plumbed through all the way.

crimsonfaith91 commented:
@enisoc Thanks for the feedback! I am writing them now. :D

crimsonfaith91 commented:
/retest

crimsonfaith91 commented:
@enisoc @janetkuo i rebased the PR and modified e2e test to use scale subresource.

crimsonfaith91 commented:
/retest

crimsonfaith91 commented:
/test pull-kubernetes-unit
/test pull-kubernetes-e2e-gce-etcd3

janetkuo commented on test/e2e/framework/statefulset_utils.go:
```diff
@@ -193,11 +194,30 @@ func getStatefulSetPodNameAtIndex(index int, ss *apps.StatefulSet) string {
 	ns := ss.Namespace
-	s.update(ns, name, func(ss *apps.StatefulSet) { *(ss.Spec.Replicas) = count })
+	newScale, _ := scaleFromStatefulSet(ss, count)
+	s.update(ns, name, func(ss *apps.StatefulSet) { *(ss.Spec.Replicas) = newScale.Spec.Replicas })
```
What you did here is:
1. create a `extensions.Scale`
1. update a StatefulSet by setting StatefulSet's `replicas` to Scale's `replicas`

This does not test `/scale` endpoint. This just sets StatefulSet's `replicas` field and is equivalent to the original code:
```go
s.update(ns, name, func(ss *apps.StatefulSet) { *(ss.Spec.Replicas) = count })
```

janetkuo replied:
The way to call an endpoint is something like this:
```go
// c is a clientset
// scale is a scale object
c.AppsV1beta2().StatefulSets(ns).Scale(scale)
```
Of course, this needs apps/v1beta2 to be enabled. 


You can take a look at how deployment's rollback endpoint is used:
https://github.com/kubernetes/kubernetes/blob/v1.8.0-alpha.2/test/e2e/workload/deployment.go#L676

crimsonfaith91 replied:
@janetkuo Noted! i didn't know `rollback` is also a subresource. thanks! :)

crimsonfaith91 replied:
I will call an endpoint and test it in the similar way.

crimsonfaith91 replied:
@janetkuo I am not sure how i should make the change. Did you mean I have to add `Scale`, `ScaleStatus` and `ScaleSpec` [here](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/kubernetes/typed/apps/v1beta2/statefulset.go)?

I want to implement something like this for testing:
```
scale := newStatefulSetScale(ss, count)
err := s.c.AppsV1beta2().StatefulSets(ns).Scale(scale)
Expect(err).NotTo(HaveOccurred())
s.update(ns, name, func(ss *apps.StatefulSet) { *(ss.Spec.Replicas) = scale.Spec.Replicas })
```
Is this a correct approach?

This will not work as `Scale` is undefined for `"k8s.io/client-go/kubernetes/typed/apps/v1beta2".StatefulSetInterface` without making the change [here](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/kubernetes/typed/apps/v1beta2/statefulset.go). However, you already have the change [here](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/api/apps/v1beta2/types.go).

It may due to we are not generating the required spec for `StatefulSetInterface`. How should i make the change if we want to have `v1beta2` disabled?

crimsonfaith91 replied:
If it is fine to enable at `master.go`, i think it requires running `hack/update-codegen.sh` to generate a new clientset, which will involve a lot of code change. Is this intended?

crimsonfaith91 replied:
Sorry for not diving deeper enough before asking the question. To answer my own question, i should:
(1) enable `v1beta2` at `master.go`
(2) enable `Scale` type at `pkg/apis/apps/types.go` and `staging/src/k8s.io/api/apps/v1beta2/types.go` by putting `// +genclient=true` annotation
(3) run `hack/update-codegen.sh` to regenerate clientset

Autogen is required as we need to enable `v1beta2` to test the subresource. I am not sure this is the right approach, but will look at this direction. :)

janetkuo replied:
We don't want to enable apps/v1beta2 by default in master.go now. I'll send a PR to https://github.com/kubernetes/test-infra/ to enable apps/v1beta2 in all e2e tests, similar to this one: https://github.com/kubernetes/test-infra/commit/caca50bce8b58833356f99f38f83282a21c3ea97

crimsonfaith91 replied:
thanks! :)

janetkuo replied:
Filed https://github.com/kubernetes/test-infra/pull/3656

crimsonfaith91 commented on pkg/registry/apps/statefulset/storage/storage.go:
```diff
@@ -17,31 +17,54 @@ limitations under the License.
 )
 
-// rest implements a RESTStorage for replication controllers against etcd
+// StatefulSetStorage includes dummy storage for StatefulSets and for Scale subresource.
```
noted!

liggitt replied:
Dummy storage? What do you mean by that?

crimsonfaith91 replied:
I implemented the framework based on replicaset. There is a similar comment for [replicaset implementation](https://github.com/kubernetes/kubernetes/blob/master/pkg/registry/extensions/replicaset/storage/storage.go#L38)

I guess the original author means `ReplicaSetStorage` encapsulates `REST`, `StatusREST` and `Registry`.

liggitt commented:
/retest

liggitt commented on pkg/registry/apps/statefulset/storage/storage.go:
```diff
@@ -92,3 +115,88 @@ var _ rest.ShortNamesProvider = &REST{}
+
+// New creates a new Scale object
+func (r *ScaleREST) New() runtime.Object {
+	return &extensions.Scale{}
```
@directxman12 @deads2k are all scale endpoints returning extensions.Scale{} forever?

crimsonfaith91 replied:
@liggitt From my understanding, as we can only have **at most one** internal version (which is `extensions` for now), the scale subresource can only be created using `extensions.Scale{}`.

DirectXMan12 replied:
so, it's kind of even worse that that.  In order to actually work, there has to be an object in the extensions category, so this storage basically doesn't matter.  I'd do what the RC storage does (return `autoscaling.Scale` from "primary" group, for use in the future, and then have the extensions storage return `extensions.Scale`)

deads2k replied:
> In order to actually work, there has to be an object in the extensions category, so this storage basically doesn't matter.

This won't do.  We should not just create another "dummy" resource that sits in the extensions group.  That is an indication that the polymorphic nature of the endpoitn has failed.

DirectXMan12 replied:
> This won't do. We should not just create another "dummy" resource that sits in the extensions group

Oh yeah, totally agree, *but* if you want this to work *now*, that's what you'd have to do.  Otherwise, this actually working with other components will have to wait.

enisoc replied:
Is the dummy object in `extensions` necessary for the actual REST API to work? Or are you just saying it's necessary for the current Scale client to work with it?

In other words, if we omit the dummy object, and do a direct `curl` to `/apis/apps/v1beta2/namespaces/default/statefulsets/blah/scale`, would that work?

deads2k replied:
> Oh yeah, totally agree, but if you want this to work now, that's what you'd have to do. Otherwise, this actually working with other components will have to wait.

It's not a valid solution to "work" now.  That isn't working and doesn't fulfill the intent of the autoscaler.  If the infrastructure needs additional effort to be made to work, then that's necessary.

If we had caught the previous dummy in time, we would have stopped it before it got out the door.

DirectXMan12 commented:
For visibility:

Currently, in order for this to work, you *need* to have `<resource>.extensions/scale`, even if the resource doesn't normally live in `extensions`.  See, for instance, the RC code.

We need to fix that in the scale client, but there are some compatibility limitations: some things accept extensions.Scale, others return accept `autoscaling.Scale` as they should, and discovery information didn't expose API group for subresources, so we have no way of knowing which is which from discovery, so we can't make a generic client @kubernetes/sig-api-machinery-misc @deads2k @liggitt .

crimsonfaith91 commented on test/e2e/framework/statefulset_utils.go:
```diff
@@ -193,11 +196,50 @@ func getStatefulSetPodNameAtIndex(index int, ss *apps.StatefulSet) string {
+	// create a new scale subresource
+	scale := newStatefulSetScale(ss, count)
+
+	// TODO: add scale endpoint to StatefulSet interface
```
@janetkuo @enisoc 
I want to have something like this: `err := s.c.AppsV1beta2().StatefulSets(ns).Scale(scale)`. It requires:
(1) implementation of `Scale` endpoint setter
(2) enable `Scale` type at `pkg/apis/apps/types.go` and `staging/src/k8s.io/api/apps/v1beta2/types.go` (done by Janet) by putting `// +genclient=true` annotation

Due to addition of annotation, i still need to run `hack/update-codegen.sh` to regenerate clientset, which should not be required. I think there is a gap between my understanding on adding an endpoint.

I had a try with PATCH. I am not sure whether this approach hits something that you guys want.

crimsonfaith91 replied:
As discussed offline, i will work on HPA and dynamic client directions.

crimsonfaith91 commented on test/e2e/autoscaling/horizontal_pod_autoscaling.go:
```diff
@@ -65,6 +65,16 @@ var _ = framework.KubeDescribe("[HPA] Horizontal pod autoscaling (scale resource
 		})
 	})
 
+	framework.KubeDescribe("[Serial] [Slow] StatefulSet", func() {
```
@janetkuo @enisoc It looks like e2e tests for scale subresource should be based on HPA's existing framework. By referring [this](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/horizontal-pod-autoscaler.md#horizontalpodautoscaler-object), `HorizontalPodAutoscalerSpec` type has a type named `SubresourceReference` referring to the scale subresource.

crimsonfaith91 replied:
I also verified existence of scale subresource for apps/v1beta2 using `kubectl proxy`.

enisoc replied:
Note that the `SubresourceReference` in that proposal was later replaced with `ScaleTargetRef` in the code. It is basically just a way to tell the HPA which other controller you want it to autoscale, by giving it the group-version-kind + name. You shouldn't need any such references in your testing, because you already know what controller you intend to scale.

crimsonfaith91 replied:
As discussed offline, i will:
(1) use `curl` to perform POST operations
(2) use dynamic client to modify the subresource
(3) remove HPA testing codes and save them for future usage

crimsonfaith91 commented:
/retest

crimsonfaith91 commented:
/retest

crimsonfaith91 commented:
@enisoc @janetkuo I added e2e test using `kubectl proxy` and `curl`. As the subresource is meant for v1beta2, i added some redundant helper testing functions such as `PauseNewPodsV1beta2` and `NewStatefulSetV1beta2`.

crimsonfaith91 commented:
/retest pull-kubernetes-verify

crimsonfaith91 commented:
/test pull-kubernetes-verify

foxish commented:
IIUC, it appears we don't need a dummy resource in extensions to use the scale endpoint directly from the extensions client, but do need it for HPA. Can we file issues tracking the steps (discovery API fix, etc) that need to happen to ensure that we can use HPA with resources that aren't in extensions? Eventually, we will want to deprecate the types in extensions, so, we're going to have to deal with this sooner or later.

crimsonfaith91 commented:
/test pull-kubernetes-e2e-gce-etcd3

crimsonfaith91 commented:
/test pull-kubernetes-cross

crimsonfaith91 commented:
/retest

crimsonfaith91 commented:
/test pull-kubernetes-e2e-gce-etcd3

crimsonfaith91 commented:
/retest

crimsonfaith91 commented:
/test pull-kubernetes-e2e-gce-etcd3

crimsonfaith91 commented on test/e2e/framework/statefulset_utils.go:
```diff
@@ -812,6 +821,65 @@ func NewStatefulSet(name, ns, governingSvcName string, replicas int32, statefulP
+// NewStatefulSetV1beta2 creates a new NGINX v1beta2 StatefulSet for testing. The StatefulSet is named name, is in namespace ns,
+// statefulPodsMounts are the mounts that will be backed by PVs. podsMounts are the mounts that are mounted directly
+// to the Pod. labels are the labels that will be usd for the StatefulSet selector.
+func NewStatefulSetV1beta2(name, ns, governingSvcName string, replicas int32, statefulPodMounts []v1.VolumeMount, podMounts []v1.VolumeMount, labels map[string]string) *appsV1beta2.StatefulSet {
```
@janetkuo The best is to have the function generic (able to accept both v1beta1 and v1beta2 types), but i am not sure what is the right way to do. An idea is to return v1 type and convert it later on.

enisoc replied:
I don't think you need to add code to handle v1beta2 at this point. You can create a v1beta1 StatefulSet as before, and it should still work to interact with its /scale subresource via the v1beta2 API. The server side does conversions to serve the same underlying set of objects under whatever version of the API the client requests.

crimsonfaith91 commented on test/e2e/apps/statefulset.go:
```diff
@@ -882,6 +889,89 @@ var _ = SIGDescribe("StatefulSet", func() {
+			framework.DeleteAllStatefulSets(c, ns)
+		})
+
+		It("should be modifiable", func() {
```
@janetkuo @enisoc i am not sure this e2e test works or not, but i am able to change `Spec.Replicas` manually (by terminal).

enisoc commented on test/e2e/apps/statefulset.go:
```diff
@@ -882,6 +889,89 @@ var _ = SIGDescribe("StatefulSet", func() {
+			_, err := c.AppsV1beta2().StatefulSets(ns).Create(ss)
+			Expect(err).NotTo(HaveOccurred())
+
+			By("starting proxy server")
```
Please add a TODO here explaining that we should use a real client instead of proxy+curl, once we have one that works for StatefulSet Scale subresource.

crimsonfaith91 replied:
sure! :)

liggitt replied:
please don't add proxy+curl to the e2e tests for this... even if you don't have client functions generated (which I'd prefer be part of this PR) you should be able to use the rest client directly (e.g. `c.AppsV1beta2().RESTClient().Get()...`)?

crimsonfaith91 replied:
This was, in fact, my initial plan. From my understanding, using the rest client directly requires enabling v1beta2 in `master.go` and generating a new clientset, which i was told that i should not do so. Do i miss something?

Refer [this](https://github.com/kubernetes/kubernetes/pull/49168#discussion_r128927356).

liggitt replied:
if you are accessing the API via `curl -> kubeproxy -> API server`, then you can get the same content via the rest client using an AbsPath() get. If you are not enabling v1beta2, then it will not be accessible via `kubectl proxy` either

liggitt replied:
Also, generation of a clientset is independent of enablement of the API version in the server. I see no reason not to generate the clientset when adding the types for a new version.

crimsonfaith91 replied:
I think i stated something incorrectly. Although v1beta2 is not enabled by default, the e2e tests for `apps/v1beta2` is enabled. I was thinking something like `c.AppsV1beta2().Scale(...)`.

I will look more into `RESTClient()`. Thanks!

enisoc commented on test/e2e/apps/statefulset.go:
```diff
@@ -882,6 +889,89 @@ var _ = SIGDescribe("StatefulSet", func() {
+				framework.Failf("Failed to start proxy server: %v", err)
+			}
+
+			ssPath := fmt.Sprintf("http://localhost:%d/apis/apps/v1beta2/namespaces/default/statefulsets/%s/scale", port, ssName)
```
You should insert the value of the `ns` variable instead of `default`.

crimsonfaith91 replied:
good point - thanks for pointing it out!

enisoc commented on test/e2e/apps/statefulset.go:
```diff
@@ -17,13 +17,18 @@ limitations under the License.
 package apps
 
 import (
+	"encoding/json"
```
Use `"k8s.io/apimachinery/pkg/util/json"` instead because it preserves k8s data type conventions (e.g. int instead of float).

crimsonfaith91 replied:
noted! i didn't know existence of the library. thanks!

enisoc commented on test/e2e/apps/statefulset.go:
```diff
@@ -882,6 +889,89 @@ var _ = SIGDescribe("StatefulSet", func() {
+			Expect((spec["replicas"]).(float64)).To(Equal(1.0))
+
+			By("changing Spec.Replicas from 1 to 9 via PUT (update)")
+			spec["replicas"] = 9.0
```
Changing to 2 would be fine. Since this is an e2e test, scaling to 9 will take extra time.

crimsonfaith91 replied:
yup!

enisoc commented on test/e2e/apps/statefulset.go:
```diff
@@ -882,6 +889,89 @@ var _ = SIGDescribe("StatefulSet", func() {
+			_, err = exec.Command(curlCmd, putArgs...).Output()
+			Expect(err).NotTo(HaveOccurred())
+
+			By("curling proxy again to verify change to Spec.Replicas")
```
We already verified the flow of data from StatefulSet.Spec -> Scale.Spec the first time we curled the subresource and expected a value of 1.

Now we should verify the opposite flow from Scale.Spec -> StatefulSet.Spec by reading the full StatefulSet object and expecting Spec.Scale = 2.

crimsonfaith91 replied:
noted. will add the test. thanks!

crimsonfaith91 replied:
@enisoc i carried out a manual testing and confirmed that the change reflected to both `StatefulSet.Spec` and `Scale.Spec`. i will look more into `RESTClient` for e2e testing.

liggitt commented on test/e2e/apps/statefulset.go:
```diff
@@ -882,6 +889,89 @@ var _ = SIGDescribe("StatefulSet", func() {
+			var scaleSubresourceSpec map[string]interface{}
+
+			By("curling proxy to obtain Spec.Replicas info")
+			body, err = exec.Command(curlCmd, args...).Output()
```
an API we can't call using our rest client doesn't seem like something we should merge. I would not expect to test this by execing curl from within an e2e test

crimsonfaith91 replied:
@liggitt There is an existing curl e2e test for kubectl: https://github.com/kubernetes/kubernetes/blob/master/test/e2e/kubectl/kubectl.go#L1475-L1495

liggitt replied:
the point of that test is to ensure a non-kube-originating network request can work through the kubectl proxy, not to work around a lack of a client for a kube api

crimsonfaith91 replied:
@liggitt I will add a TODO stating using a real client instead of proxy+curl (when having one working for StatefulSet Scale subresource).

crimsonfaith91 commented on test/e2e/apps/statefulset.go:
```diff
@@ -882,6 +889,89 @@ var _ = SIGDescribe("StatefulSet", func() {
 		})
 	})
+
+	framework.KubeDescribe("V1beta2 StatefulSet Scale Subresource [Feature:StatefulSet]", func() {
```
Ya, we can put under "Basic StatefulSet functionality". The reason i separate it out is due to it is a v1beta2 test. As confirmed by you that  a v1beta1 StatefulSet is viable due to server side conversion (internal version), there is no need anymore for separating out the test.

crimsonfaith91 commented:
I will add a TODO stating using a real client instead of proxy+curl (when having one working for StatefulSet Scale subresource).

liggitt commented on pkg/registry/apps/statefulset/storage/storage.go:
```diff
@@ -92,3 +115,88 @@ var _ rest.ShortNamesProvider = &REST{}
+func (r *ScaleREST) Get(ctx genericapirequest.Context, name string, options *metav1.GetOptions) (runtime.Object, error) {
+	ss, err := r.registry.GetStatefulSet(ctx, name, options)
+	if err != nil {
+		return nil, errors.NewNotFound(extensions.Resource("statefulsets/scale"), name)
```
this should only return a NotFound error if we got a NotFound error looking up the stateful set, right?

crimsonfaith91 replied:
yup! i will change the codes accordingly.

liggitt commented on pkg/registry/apps/statefulset/storage/storage.go:
```diff
@@ -92,3 +115,88 @@ var _ rest.ShortNamesProvider = &REST{}
+func (r *ScaleREST) Update(ctx genericapirequest.Context, name string, objInfo rest.UpdatedObjectInfo) (runtime.Object, bool, error) {
+	ss, err := r.registry.GetStatefulSet(ctx, name, &metav1.GetOptions{})
+	if err != nil {
+		return nil, false, errors.NewNotFound(extensions.Resource("statefulsets/scale"), name)
```
same here, I'd only expect NotFound errors to cause us to return NotFound errors

crimsonfaith91 commented:
@liggitt I had a look at RESTClient library. I am not sure whether my approach below is something that you meant:
```
ssPath := fmt.Sprintf("/namespaces/%s/statefulsets/", ns)
scalePath := fmt.Sprintf("%s/%s/scale/", ssPath, ssName)

By("creating scale subresource using POST")
_, err = f.ClientSet.Core().RESTClient().Post().AbsPath(ssPath).DoRaw()
if err != nil {
    framework.Failf("Failed to create scale subresource: %v", err)
}
```

crimsonfaith91 commented on test/e2e/apps/statefulset.go:
```diff
@@ -843,6 +847,75 @@ var _ = SIGDescribe("StatefulSet", func() {
 			}, framework.StatefulPodTimeout, 2*time.Second).Should(BeNil())
 		})
+
+		It("should have a working scale subresource", func() {
```
@liggitt I am not sure whether i catch your meaning. This e2e test does not work yet as the scale subresource cannot be found.

What i want is to obtain a JSON blob located at:
(1) for StatefulSet: e.g., `http://localhost:8001/apis/apps/v1beta2/namespaces/default/statefulsets/`
(2) for StatefulSet's scale subresource: e.g., `http://localhost:8001/apis/apps/v1beta2/namespaces/default/statefulsets/web/scale`

I am able to test it manually using `kubectl proxy` and `curl`. The subresource `Spec.Replicas` attribute can be modified, and the change will be reflected to the StatefulSet `Spec.Replicas` attribute.

I am not sure whether `RESTClient` is the suitable tool for this. I will look more into the library, but do you have any recommended person that i can ask for help with regarding this? Thanks!

liggitt replied:
> I am able to test it manually using kubectl proxy and curl

all `kubectl proxy` does is add credentials from your local kubeconfig file to requests made to `http://localhost:8001/...` and pass the request along to the API server. if this is working for you, it must mean you have `apps/v1beta2` enabled on your API server, and could make the same request directly against the API server, without `kubectl proxy` being involved at all.

if all you want to do in the e2e is to submit a request to an API endpoint, there's no need to use curl, or to run `kubectl proxy`.

I'd expect the e2e test to do something more like this:
```
			By("getting scale subresource")
			scale := &extensionsv1beta1.Scale{}
			err := c.AppsV1beta2().RESTClient().Get().AbsPath("/apis/apps/v1beta1").Namespace(ns).Resource("statefulsets").Name(ssName).SubResource("scale").Do().Into(scale)
			if err != nil {
				framework.Failf("Failed to get scale subresource: %v", err)
			}
			Expect(scale.Spec.Replicas).To(Equal(1))
			Expect(scale.Status.Replicas).To(Equal(1))

			By("updating a scale subresource")
			// unconditionally update to 2 replicas
			scale.ResourceVersion = ""
			scale.Spec.Replicas = 2
			scaleResult := &extensionsv1beta1.Scale{}
			err := c.AppsV1beta2().RESTClient().Put().AbsPath("/apis/apps/v1beta1").Namespace(ns).Resource("statefulsets").Name(ssName).SubResource("scale").Body(scale).Do().Into(scaleResult)
			if err != nil {
				framework.Failf("Failed to put scale subresource: %v", err)
			}
			Expect(scaleResult.Spec.Replicas).To(Equal(2))

			By("verifying the stateful set spec was modified")
			ss, err := c.AppsV1beta2().StatefulSets(ns).Get(ssName, metav1.GetOptions{})
			if err != nil {
				framework.Failf("Failed to get stateful set resource: %v", err)
			}
			Expect(ss.Spec.Replicas).To(Equal(2))
```

crimsonfaith91 replied:
@liggitt Thank you very much for your detailed explanation! i will dive deep into each of the components carefully, and get back to you tomorrow. Thanks!

crimsonfaith91 commented:
@enisoc @janetkuo I have some difficulties writing e2e test for the scale subresource. I am able to test that the feature is working via manual testing using `kubectl proxy` and `curl`. I am not sure whether the e2e test is really needed now, or we can add a TODO for future when we have a generic Scale client. At that moment, we can add the same test for other API objects such as `Deployment` too.

crimsonfaith91 commented:
@liggitt `v1beta1` client will not work as there is no appropriate `Get()` function for scale subresource.
Internal version has the function at `ScaleExpansion`, but not `v1beta1` version: https://github.com/kubernetes/kubernetes/blob/master/pkg/client/clientset_generated/internalclientset/typed/extensions/internalversion/scale_expansion.go#L32-L47

liggitt commented:
You have to add it to the scale expansion. The expansion changes are manual

crimsonfaith91 commented:
@liggitt Yes, i added it before posting my finding. i spent some time resolving another issue due to `protobuf marshalling` - pinged you on Slack if you happened to encounter it before.

crimsonfaith91 commented:
@liggitt The test should pass now. :)

crimsonfaith91 commented on test/e2e/apps/statefulset.go:
```diff
@@ -850,44 +850,51 @@ var _ = SIGDescribe("StatefulSet", func() {
-			scale := &extensions.Scale{}
-			err = c.AppsV1beta1().RESTClient().Get().AbsPath("/apis/apps/v1beta2").Namespace(ns).Resource("statefulsets").Name(ssName).SubResource("scale").Do().Into(scale)
+			
+			err = c.AppsV1beta2().RESTClient().Get().AbsPath("/apis/apps/v1beta2").Namespace(ns).Resource("statefulsets").Name(ssName).SubResource("scale").Do().Into(scale)
```
@liggitt The test passes on local cluster, but not in submit queue. It is most probably due to client version incompatibility.

It should be hard moving forward - i wish to deliver gratitude for your patient help and detailed explanation. Thanks!

crimsonfaith91 replied:
@liggitt The test passes on local cluster, but not on submit queue. It most probably is due to client version incompatibility. i have some ideas on resolving it.

It should not be hard moving forward - i wish to deliver gratitude for your patient help and detailed explanation. Thanks!

liggitt commented on pkg/registry/apps/rest/storage_apps.go:
```diff
@@ -86,9 +86,10 @@ func (p RESTStorageProvider) v1beta2Storage(apiResourceConfigSource serverstorag
+		statefulSetStorage := statefulsetstore.NewStorage(restOptionsGetter)
+		storage["statefulsets"] = statefulSetStorage.StatefulSet
+		storage["statefulsets/status"] = statefulSetStorage.Status
+		storage["statefulsets/scale"] = statefulSetStorage.Scale
```
is there a reason not to add `statefulsets/scale` to v1beta1?

crimsonfaith91 replied:
@kow3ns

kow3ns replied:
@liggitt we want to advance v1beta2 to v1. So we are putting our effort there. There is no reason to not add a scale subresource to v1beta1 (it would be backward compatible). Do you have a strong reason or concern that would motivate adding a scale subresource to v1beta1?

liggitt replied:
consistency is preferred if there's no downside

crimsonfaith91 commented:
/retest

crimsonfaith91 commented:
/retest

crimsonfaith91 commented:
/retest

krzyzacy commented:
/test pull-kubernetes-e2e-gce-etcd3

crimsonfaith91 commented:
/test pull-kubernetes-e2e-gce-etcd3

crimsonfaith91 commented:
/test pull-kubernetes-e2e-kops-aws

crimsonfaith91 commented:
/retest

enisoc commented on test/e2e/apps/statefulset.go:
```diff
@@ -843,6 +847,52 @@ var _ = SIGDescribe("StatefulSet", func() {
+			sst.WaitForRunningAndReady(*ss.Spec.Replicas, ss)
+			ss = sst.WaitForStatus(ss)
+
+			By("posting a scale subresource")
```
What is the benefit of this step in the test? It seems to test the same thing as "updating a scale subresource".

crimsonfaith91 replied:
i think we should have a step that explicitly creates the scale subresource. i will double check whether this step is optional or not. If it is, i will remove the step.

crimsonfaith91 replied:
Ya, posting a scale subresource is not necessary.

crimsonfaith91 replied:
@enisoc @kow3ns I verified that posting scale subresource right at the beginning is not needed. Thanks for the feedback!

enisoc commented on test/e2e/apps/statefulset.go:
```diff
@@ -843,6 +847,52 @@ var _ = SIGDescribe("StatefulSet", func() {
+			if err != nil {
+				framework.Failf("Failed to get scale subresource: %v", err)
+			}
+			fmt.Println(scale)
```
Is this temporary? If you think it will be needed long-term to debug test failures, it should be a `glog.Infof()` so it prints the filename and line number. If you only needed it for developing the test, make sure to remove any temporary print statements before submitting so we don't clutter the logs.

crimsonfaith91 replied:
Yes, this is temporary and will be removed afterwards. :)

kow3ns replied:
If you need to log use framework.Log, not fmt.Println. Also, you might want to make that Log statement a little more descriptive.

crimsonfaith91 commented:
The reason submit queue test `pull-kubernetes-e2e-gce-etcd3` failed is due to not registering `v1beta2` storage:
```
W0801 20:56:39.586156       5 genericapiserver.go:321] Skipping API apps/v1beta2 because it has no resources.
```
`apps/v1beta2` is skipped due to `VersionedResourcesStorageMap` does not have the version as the key.

IMO there are two ways to resolve this:
(1) register the `v1beta2` storage to the API server
(2) register scale subresource for `v1beta1` and test with `v1beta1`

I will go ahead with (1) by understanding API server and storage codes. Registering the subresource to `v1beta1` just for testing purpose does not sound like a good idea.

crimsonfaith91 commented:
Yes, `fmt.Println()` is definitely a bad way to add the logging message. It
is meant for debugging purpose, and will be removed afterwards.

On Tue, Aug 1, 2017 at 6:12 PM, Kenneth Owens <notifications@github.com>
wrote:

> *@kow3ns* commented on this pull request.
> ------------------------------
>
> In test/e2e/apps/statefulset.go
> <https://github.com/kubernetes/kubernetes/pull/49168#discussion_r130766443>
> :
>
> > +			By("posting a scale subresource")
> +			scale := framework.NewStatefulSetScale(ss)
> +			scaleResult := &appsV1beta2.Scale{}
> +			err = c.AppsV1beta2().RESTClient().Put().AbsPath("/apis/apps/v1beta2").Namespace(ns).Resource("statefulsets").Name(ssName).SubResource("scale").Body(scale).Do().Into(scaleResult)
> +			if err != nil {
> +				framework.Failf("Failed to put scale subresource: %v", err)
> +			}
> +			Expect(scaleResult.Spec.Replicas).To(Equal(int32(1)))
> +
> +			By("getting scale subresource")
> +
> +			err = c.AppsV1beta2().RESTClient().Get().AbsPath("/apis/apps/v1beta2").Namespace(ns).Resource("statefulsets").Name(ssName).SubResource("scale").Do().Into(scale)
> +			if err != nil {
> +				framework.Failf("Failed to get scale subresource: %v", err)
> +			}
> +			fmt.Println(scale)
>
> If you need to log use framework.Log, not fmt.Println. Also, you might
> want to make that Log statement a little more descriptive.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/49168#discussion_r130766443>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ANRtXo3QyDyCccaJnk9N_w0nARNFgjJcks5sT80CgaJpZM4OcG7x>
> .
>

crimsonfaith91 commented:
/test pull-kubernetes-unit

crimsonfaith91 commented:
/retest

crimsonfaith91 commented:
/retest

crimsonfaith91 commented:
/retest

crimsonfaith91 commented:
/test pull-kubernetes-unit

liggitt commented on pkg/registry/apps/rest/storage_apps.go:
```diff
@@ -63,9 +63,13 @@ func (p RESTStorageProvider) v1beta1Storage(apiResourceConfigSource serverstorag
-		storage["statefulsets/status"] = statefulsetStatusStorage
+		// enable scale subresource for v1beta1 for e2e testing
+		// we are not able to test v1beta2 as v1beta2 is disabled at master.go
+		// TODO (juntee): remove the subresource before 1.8 release
```
I don't see a reason to remove this before 1.8... why wouldn't we want the scale subresource in v1beta1 as well?

crimsonfaith91 replied:
IMO there is no reason to add a new subresource to a version to be retired soon. The only reason we add the /scale for v1beta1 is for e2e testing purpose. If having integration test in this PR is better, we should not add /scale to v1beta1.
@kow3ns @foxish

crimsonfaith91 replied:
In this case, i prefer spending more time investigating how to make e2e test works for v1beta2.

crimsonfaith91 replied:
@liggitt @kow3ns After discussion, `/scale` will NOT be remove for `v1beta1` as there is no maintenance cost.

I plan to:
(1) shift e2e test from v1beta1 to v1beta2 BEFORE code freeze
(2) replace e2e test with integration test sometime AFTER code freeze and BEFORE 1.8 release

Tell me if there is a better way. Thanks!

liggitt replied:
> I plan to:
> (1) change e2e test's version from v1beta1 to v1beta2 BEFORE code freeze
> (2) replace e2e test with integration test sometime AFTER 1.8 release

SGTM. if the switch to integration test happens before code freeze, that works as well.

crimsonfaith91 replied:
Thanks for confirming! I opened issue #50109 for reference.

kow3ns commented:
@janetkuo are you satisfied that @crimsonfaith91 has addressed your review?

crimsonfaith91 commented:
@liggitt If everything looks good, do you want to approve this PR? Thanks!

liggitt commented:
/approve, though still needs @kubernetes/api-approvers and LGTM

crimsonfaith91 commented:
@liggitt Thanks a lot! :)

kow3ns commented:
I will /lgtm

kow3ns commented:
/lgtm

crimsonfaith91 commented:
@kow3ns Many thanks! :)

crimsonfaith91 commented:
/assign @erictune

smarterclayton commented on test/e2e/apps/statefulset.go:
```diff
@@ -47,6 +48,8 @@ const (
 	readTimeout = 60 * time.Second
 )
 
+var proxyRegexp = regexp.MustCompile("Starting to serve on 127.0.0.1:([0-9]+)")
```
This looks no longer needed?

crimsonfaith91 replied:
@smarterclayton Thanks for reviewing! Yes, this line is unnecessary. :)

smarterclayton commented:
API approved.  One minor comment, you can do that in a follow up.

/approve

crimsonfaith91 commented:
/retest

crimsonfaith91 commented:
/retest

kow3ns commented:
/lgtm
